apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/RetriableSpannerTransaction.cs

```diff
@@ -0,0 +1,374 @@
+        {
+            RetryAsync(CancellationToken.None, timeoutSeconds).WaitWithUnwrappedExceptions();
+        }
+
+        internal async Task RetryAsync(CancellationToken cancellationToken, int timeoutSeconds = MAX_TIMEOUT_SECONDS)
+        {
+            while (true)
+            {
+                _retryCount++;
+                Session = await (Session?.WithFreshTransactionOrNewAsync(SpannerConnection.s_readWriteTransactionOptions, cancellationToken) ?? SpannerConnection.AcquireReadWriteSessionAsync(cancellationToken)).ConfigureAwait(false);
+                try
+                {
+                    foreach (IRetriableStatement statement in _retriableStatements)
+                    {
+                        await statement.Retry(this, cancellationToken, timeoutSeconds).ConfigureAwait(false);
+                    }
+                    break;
+                } catch (SpannerAbortedDueToConcurrentModificationException e)
+                {
+                    // Retry failed because of a concurrent modification.
+                    throw e;
+                } catch (SpannerException e) when (e.ErrorCode == ErrorCode.Aborted)
+                {
+                    // Ignore and retry.
+                    if (_retryCount >= MAX_RETRIES)
+                    {
+                        throw new SpannerException(ErrorCode.Aborted, "Transaction was aborted because it aborted and retried too many times");
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Commits the database transaction asynchronously.
+        /// </summary>
+        /// <param name="cancellationToken">A cancellation token used for this task.</param>
+        /// <returns>Returns the UTC timestamp when the data was written to the database.</returns>
+        public async override Task<DateTime> CommitAsync(CancellationToken cancellationToken = default)
+        {
+            while (true)
+            {
+                try
+                {
+                    return await base.CommitAsync(cancellationToken).ConfigureAwait(false);
+                }
+                catch (SpannerException e) when (e.ErrorCode == ErrorCode.Aborted)
+                {
+                    await RetryAsync(cancellationToken).ConfigureAwait(false);
+                }
+            }
+        }
+
+        /// <inheritdoc />
+        public override void Commit()
+        {
+            while (true)
+            {
+                try
+                {
+                    base.Commit();
+                }
+                catch (SpannerException e) when (e.ErrorCode == ErrorCode.Aborted)
+                {
+                    Retry();
+                }
+            }
+        }
+
+        Task<int> ISpannerTransaction.ExecuteMutationsAsync(List<Mutation> mutations, CancellationToken cancellationToken, int timeoutSeconds)
+        {
+            return DoExecuteMutationsAsync(mutations, cancellationToken, timeoutSeconds);
```

Comment:
This is the problem right here. These mutations might be based on data that is stale when they get to be executed as part of the commit.
Imagine a read goes through, and we create a mutation based on that read, and add it to the transaction mutations. Then we have to retry because of some other operation was aborted. While we are retrying, someone updates the value we based the mutation on, now our mutation has become stale. Even if we retry the original read in which we based the mutation, the mutation itself won't be updated according to the newly read values, which would make the mutation stale. We would effectively be allowing through mutations that are based on invalid read locks (the original one that we had to retry).

We explored this approach and found it wasn't possible, you really need to take a delegate in that contains the whole sets of intrstuctions that need to be retried.

Reply:
No, the retry logic is checking for exactly that possibility, and if it detects it, it will abort the transaction after all. Consider the following (pseudo-code) example:

```
SELECT Id, Name FROM Singers WHERE Id=1;
-- The following statement is not valid SQL, but hopefully it's clear what's meant.
CREATE UPDATE_MUTATION Singer (Id, Name) VALUES (1, 'NewName');
COMMIT;
```

Now let's assume that another transaction updates the name of the same Singer to 'SomeOtherName' at the same time and commits. The above transaction will be aborted by Cloud Spanner. The `Aborted` error is in this example returned during the COMMIT call (but any statement could return `Aborted`) and the transaction will have to execute a retry.

The `RetriableSpannerTransaction` keeps track of all statements that have been executed during the transaction and the results that it received. When a retry is needed, it starts a new transaction and executes all statements again on the new transaction and compares the results that it receives during the retry with the results that it received during the initial transaction. If they match, the transaction is allowed to proceed as if nothing happened. If during the retry it receives different data, the retry will abort and the `Aborted` error will be propagated to the client application.

In the above example the first query [will be returned](https://github.com/googleapis/google-cloud-dotnet/blob/c99cc806b78137f032cc5f0674101236b579f6e7/apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/RetriableSpannerTransaction.cs#L370) as a `SpannerDataReaderWithChecksum`. That reader keeps track of a running checksum of the data it has already seen, and if a retry is needed, it will compare the checksum from the original transaction with the checksum during the retry. In this example the checksums will not match, and the transaction will abort as it should.
If however the specific Singer row had not been updated and the transaction was aborted for some other reason, the retry will succeed.

Reply:
For some more context: This retry strategy is also what is used in the [Cloud Spanner JDBC driver](https://github.com/googleapis/java-spanner/blob/master/google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/ChecksumResultSet.java).

Reply:
Then I would argue that this strategy is way more restrictive than the one we have right now in that it will result in less successful retries. Take for instance the checksum data reader, it requires rows to be always returned in exactly the same order, I'm not sure this is guarenteed by Spanner. There's also a requirement for updated number of rows and errors even to be the same every time, just because you can only hold onto the mutations as they were initially. So in a sense this goes further than guaranteeing non-stale mutations, it enforces that the transaction will end up committing only if it can do so exactly the same as it would have the first time. This is in my opinion way more restrictive than usually expected for a retriable transaction.

Also, I wouldn't say that mutations are not going stale. They might not be stale with respect to the pure Spanner data, but they can go "semantically" stale. Say you have several mutations that store DateTime.Now. Depending on how many retries you get in between adding these mutations, the relation between those values will be wildly different and probably not meaningful. Or for instance, we would need to remember to never ever expose the Session object in the SpannerTransaction so that no one can base a mutation on that.

And the checksum overhead, I'd say that's not trivial either in terms of memory use or of computational complexity.

I don't want to sound overly negative or far fetched. But I think this is significant work that comes with a maintenance burden and I honestly don't see it adds much to what we already have. It's only just a little bit more like ADO.NET but it's not even that because ADO.NET does not define a retriable transaction, you still wouldn't be able to use the SpannerConnection as a DbConnection and get to this feature, same as what we have now.
Anyway, those are my two cents, @skuruppu and @jonskeet might think differently.